File: core/java/src/main/java/hara/lang/protocol/ILookup.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;
import java.util.Iterator;
import java.util.Map;

@HaraProtocolBinding(
    namespace = "std.protocol.ilookup", name = "ILookup", parents = {"IFind"})
public interface ILookup<K, V> extends IFind<K, Map.Entry<K, V>> {

  Iterator<K> keys();

  default V lookup(K key) {
    return lookup(key, null);
  }

  @HaraMethod(value = "lookup", arity = -1, variadic = true)
  default V lookup(K key, V notFound) {
    Map.Entry<K, V> ret = find(key);
    return (ret == null) ? notFound : ret.getValue();
  }

  Iterator<V> vals();
}

File: core/rust/src/lang/protocol/ilookup.rs
use super::IFind;
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.ilookup",
    name = "ILookup",
    parents = ["IFind"]
)]
pub trait ILookup<K, V>: IFind<K, Output = (K, V)>
where
    K: Clone,
    V: Clone,
{
    type Keys: Iterator<Item = K>;
    type Values: Iterator<Item = V>;
    fn keys(&self) -> Self::Keys;
    fn vals(&self) -> Self::Values;
    #[hara_method(
        value = "lookup",
        arity = -1,
        variadic = true,
        min_arity = 2,
        max_arity = 3,
        whole_wasm
    )]
    fn lookup(&self, key: &K) -> Option<V> {
        self.find(key).map(|(_, value)| value)
    }
    fn lookup_or(&self, key: &K, not_found: V) -> V {
        self.lookup(key).unwrap_or(not_found)
    }
}
